### PR TITLE
[l2-load-balancer][ingress-nginx] The new L2 Load Balancer

### PR DESCRIPTION
**Pull request by Beastlex**
## Description

The new module `l2-load-balancer` and a new `IngressNginxController` inlet `L2LoadBalancer`.

## Why do we need it, and what problem does it solve?

In the current implementation, LoadBalancers based on MetalLB L2 are not fault-tolerant. We are introducing the ability to deploy LoadBalancer services (for MetalLB) across all nodes selected via nodeSelector. This approach prevents the loss of all traffic when one of the MetalLB speakers becomes unavailable.

## Checklist
- [x] The code is covered by unit tests.
- [x] e2e tests passed.
- [x] Documentation updated according to the changes.
- [x] Changes were tested in the Kubernetes cluster manually.

## Changelog entries

```changes
section: l2-load-balancer
type: feature
summary: The new module for redundant L2 load-balancing and a new IngressNginxController inlet L2LoadBalancer.
impact_level: default
```

### CODE REVIEW

File: ee/modules/380-metallb/images/artifact/werf.inc.yaml [hunk 1]
```diff
@@ -11,6 +17,7 @@ shell:
     # https://github.com/metallb/metallb/pull/2004
     # Remove cherry-pick when this commit is included in the release
     - git cherry-pick 667ab7e93d1e6465d5b6d343b73b3c7b1edd5015
+    - test -d /patches/ && for patchfile in /patches/*.patch ; do git apply ${patchfile} ; done
```
Comment (apolovov) [inline ee/modules/380-metallb/images/artifact/werf.inc.yaml:20]:
```
find /patches -name '*.patch' | xargs git apply --verbose
```

File: modules/402-ingress-nginx/crds/ingress-nginx.yaml [hunk 2]
```diff
@@ -43,6 +43,7 @@ spec:
                     The way traffic goes to cluster from the outer network.
                     * `LoadBalancer` — Ingress controller is deployed and the service of `LoadBalancer` type is provisioned.
                     * `LoadBalancerWithProxyProtocol` — Ingress controller is deployed and the service of `LoadBalancer` type is provisioned. Ingress controller uses proxy-protocol to get a real IP of the client.
+                    * `L2LoadBalancer` - Ingress controller is deployed and the services of `LoadBalancer` (MetalLB) type is provisioned in resilient mode.
```
Comment (apolovov) [inline modules/402-ingress-nginx/crds/ingress-nginx.yaml:46]: ?

File: modules/402-ingress-nginx/crds/ingress-nginx.yaml [hunk 3]
```diff
@@ -96,6 +97,22 @@ spec:
...
                     Requires a controller of 1.9 version.
+                l2LoadBalancer:
+                  type: object
+                  description: |
+                    A section of parameters of the `LoadBalancer` inlet.
+                  properties:
+                    addressPool:
+                      type: string
+                      description: |
+                        Name of MetalLB pool.
```
Comment (apolovov) [inline modules/402-ingress-nginx/crds/ingress-nginx.yaml:108]:
```suggestion
                        Name of MetalLB address pool.
```

File: modules/402-ingress-nginx/hooks/get_ingress_controllers.go [hunk 4]
```diff
@@ -39,9 +47,30 @@ var _ = sdk.RegisterFunc(&go_hook.HookConfig{
 			Kind:       "IngressNginxController",
 			FilterFunc: applyControllerFilter,
 		},
+		{
+			Name:                         "nodes",
+			ApiVersion:                   "v1",
+			Kind:                         "Node",
+			WaitForSynchronization:       pointer.Bool(false),
+			ExecuteHookOnSynchronization: pointer.Bool(false),
```
Comment (apolovov) [inline modules/402-ingress-nginx/hooks/get_ingress_controllers.go:55]: why not?

File: ee/modules/030-cloud-provider-vcd/candi [hunk 5]
```diff
@@ -1 +0,0 @@
-/deckhouse/candi/cloud-providers/vcd
```
Comment (apolovov) [inline ee/modules/030-cloud-provider-vcd/candi:1]: ?

File: modules/402-ingress-nginx/crds/ingress-nginx.yaml [hunk 6]
```diff
@@ -96,6 +97,22 @@ spec:
                     Enables the annotation validation feature.
 
                     Requires a controller of 1.9 version.
+                l2LoadBalancer:
```
Comment (apolovov) [inline modules/402-ingress-nginx/crds/ingress-nginx.yaml:100]: move closer to `loadBalancer`

File: modules/402-ingress-nginx/hooks/get_ingress_controllers.go (hunk 4 above)
Comment (apolovov) [inline modules/402-ingress-nginx/hooks/get_ingress_controllers.go:55]:
```suggestion
```

File: modules/402-ingress-nginx/hooks/get_ingress_controllers.go [hunk 7]
```diff
@@ -144,6 +174,41 @@ func setInternalValues(input *go_hook.HookInput) error {
...
 			version = defaultControllerVersion // it's used only for metrics
 		}
+
+		inlet, found, err := unstructured.NestedString(controller.Spec, "inlet")
+		if err != nil {
+			return fmt.Errorf("cannot get inlet type from ingress controller spec: %v", err)
+		}
+		if found && inlet == "L2LoadBalancer" {
+			var nodeSelectorLabels = labels.Everything()
+			if nodeSelector, exists := controller.Spec["nodeSelector"]; exists {
```
Comment (apolovov) [inline modules/402-ingress-nginx/hooks/get_ingress_controllers.go:184]:
```suggestion
			if nodeSelector, ok := controller.Spec["nodeSelector"]; ok {
```

File: modules/402-ingress-nginx/hooks/get_ingress_controllers.go [hunk 8]
```diff
@@ -144,6 +174,41 @@ func setInternalValues(input *go_hook.HookInput) error {
...
+				nodeSelectorMapInterface := nodeSelector.(map[string]interface{})
+				nodeSelectorMapStr := make(map[string]string)
+				for key, value := range nodeSelectorMapInterface {
+					strValue := value.(string)
+					nodeSelectorMapStr[key] = strValue
+				}
+				nodeSelectorLabels = labels.Set(nodeSelectorMapStr).AsSelector()
+			}
```
Comment (apolovov) [inline modules/402-ingress-nginx/hooks/get_ingress_controllers.go:192]:
```suggestion
			}
			nodeSelectorLabels = labels.Set(nodeSelectorMapStr).AsSelector()
```
Maybe it is the solution for empty nodeSelector?

File: modules/402-ingress-nginx/hooks/get_ingress_controllers.go [hunk 9]
```diff
@@ -144,6 +174,41 @@ func setInternalValues(input *go_hook.HookInput) error {
...
+
+			nodes := make([]map[string]string, 0, len(controllersFilterResult))
```
Comment (apolovov) [inline modules/402-ingress-nginx/hooks/get_ingress_controllers.go:194]:
```suggestion
			nodes := make([]map[string]string, 0, len(nodesFilterResult))
```

Comment (apolovov) [inline modules/402-ingress-nginx/hooks/get_ingress_controllers.go:194]:
```suggestion
             // nodes["name"] = "kube-front-0"
			nodes := make([]map[string]string, 0, len(controllersFilterResult))
```

File: modules/402-ingress-nginx/hooks/get_ingress_controllers.go [hunk 10]
```diff
@@ -144,6 +174,41 @@ func setInternalValues(input *go_hook.HookInput) error {
...
+				nodeInfo := nl.(NodeLabelsInfo)
+
+				if nodeSelectorLabels.Matches(labels.Set(nodeInfo.Labels)) {
+					nodes = append(
+						nodes,
+						map[string]string{"name": nodeInfo.Name},
+					)
+				}
+			}
+			nestedL2Balancer := make(map[string]interface{})
```
Comment (apolovov) [inline modules/402-ingress-nginx/hooks/get_ingress_controllers.go:207]: nested?

File: modules/402-ingress-nginx/crds/ingress-nginx.yaml (hunk 2 above)
Comment (apolovov) [inline modules/402-ingress-nginx/crds/ingress-nginx.yaml:46]: Also, we have to enrich the [examples](https://deckhouse.io/documentation/v1/modules/402-ingress-nginx/examples.html) page.

File: ee/modules/380-metallb/images/artifact/patches/001-l2-speker-preffered-node.patch [hunk 11]
```diff
@@ -0,0 +1,38 @@
+diff --git a/speaker/layer2_controller.go b/speaker/layer2_controller.go
```
Comment (apolovov) [inline ee/modules/380-metallb/images/artifact/patches/001-l2-speker-preffered-node.patch:1]: NB for reviewers! For now, the new inlet rely only on nodeSelector to detect frontend-nodes, not tolerations. Not sure we need them.

File: modules/402-ingress-nginx/hooks/get_ingress_controllers.go [hunk 12]
```diff
@@ -144,6 +174,41 @@ func setInternalValues(input *go_hook.HookInput) error {
 			// controllerVersion should be absent if not specified explicitly
 			version = defaultControllerVersion // it's used only for metrics
 		}
+
+		inlet, found, err := unstructured.NestedString(controller.Spec, "inlet")
+		if err != nil {
+			return fmt.Errorf("cannot get inlet type from ingress controller spec: %v", err)
+		}
+		if found && inlet == "L2LoadBalancer" {
+			var nodeSelectorLabels = labels.Everything()
```
Comment (apolovov) [inline modules/402-ingress-nginx/hooks/get_ingress_controllers.go:183]:
```suggestion
			var nodeSelectorLabels = labels.Everything() // by default, the selector is "every node"
```

File: modules/402-ingress-nginx/hooks/get_ingress_controllers.go (hunk 10 above)
Comment (apolovov) [reply modules/402-ingress-nginx/hooks/get_ingress_controllers.go:207]:
```suggestion
			inletL2LoadBalancer := make(map[string]interface{})
```

File: modules/402-ingress-nginx/hooks/get_ingress_controllers_test.go [hunk 13]
```diff
@@ -80,6 +80,7 @@ spec:
   },
```
Comment (apolovov) [inline modules/402-ingress-nginx/hooks/get_ingress_controllers_test.go:80]: Add test cases pls.

File: modules/402-ingress-nginx/templates/controller/service-l2-load-balancer.yaml [hunk 14]
```diff
@@ -0,0 +1,40 @@
+{{- $context := . }}
+{{- range $crd := $context.Values.ingressNginx.internal.ingressControllers }}
+  {{- if eq $crd.spec.inlet "L2LoadBalancer" }}
+    {{ $addressPool := $crd.spec.l2LoadBalancer.addressPool }}
+    {{ $sourceRanges := $crd.spec.l2LoadBalancer.sourceRanges }}
+{{- range $index, $node := $crd.spec.l2LoadBalancer.nodes }}
```
Comment (apolovov) [inline modules/402-ingress-nginx/templates/controller/service-l2-load-balancer.yaml:6]:
```suggestion
    {{- range $index, $node := $crd.spec.l2LoadBalancer.nodes }}
```

File: modules/402-ingress-nginx/templates/controller/service-l2-load-balancer.yaml [hunk 15]
```diff
@@ -0,0 +1,40 @@
...
+      protocol: TCP
+    - name: https
+      port: 443
+      targetPort: 443
+      protocol: TCP
+  selector:
+    app: controller
+    name: {{ $crd.name }}
+  {{- end }}
+{{- end }}
```
Comment (apolovov) [inline modules/402-ingress-nginx/templates/controller/service-l2-load-balancer.yaml:39]:
```suggestion
    {{- end }}
  {{- end }}
```

File: modules/402-ingress-nginx/templates/controller/service-l2-load-balancer.yaml [hunk 16]
```diff
@@ -0,0 +1,40 @@
+{{- $context := . }}
```
Comment (apolovov) [inline modules/402-ingress-nginx/templates/controller/service-l2-load-balancer.yaml:1]: add a template_test pls.

File: modules/402-ingress-nginx/crds/doc-ru-ingress-nginx.yaml [hunk 17]
```diff
@@ -18,6 +18,9 @@ spec:
                     Способ поступления трафика из внешнего мира:
                     * `LoadBalancer` — устанавливается Ingress-контроллер и заказывается сервис с типом `LoadBalancer`;
                     * `LoadBalancerWithProxyProtocol` — устанавливается Ingress-контроллер и заказывается сервис с типом `LoadBalancer`. Ingress-контроллер использует proxy-protocol для получения настоящего IP-адреса клиента;
+                    * `L2LoadBalancer` - cпециальный inlet для baremetal инсталляций без поддержки BGP балансировщиков. Контроллер устанавливается на каждую фронтенд-ноду, при этом каждая фронтенд-нода имеет по соответствующему сервису с типом LoadBalancer под управлением MetalLB в режиме L2. В результате, один IngressNginxController имеет N (по количеству frontend-нод) публичных отказоустойчивых IP-адресов и каждая frontend-нода в штатоном режиме участвует в балансировке трафика.
+
+                      **Внимание!** Необходимо предварительно настроить [MetalLB](/documentation/v1/modules/380-metallb/) в режиме L2 для использования данного inlet-типа.
```
[the remark this answers is not in the document]
Comment (apolovov) [reply modules/402-ingress-nginx/crds/doc-ru-ingress-nginx.yaml:23]: Inlet is available only in EE.

File: modules/402-ingress-nginx/docs/EXAMPLES.md [hunk 18]
```diff
@@ -128,4 +128,41 @@ metallb:
       value: frontend
 ```
 
+## An example for Bare metal (MetalLB Load Balancer in L2 Mode)
```
Comment (apolovov) [inline modules/402-ingress-nginx/docs/EXAMPLES.md:131]: Add a BGP note in example above.

File: modules/402-ingress-nginx/docs/EXAMPLES_RU.md [hunk 19]
```diff
@@ -128,4 +128,41 @@ metallb:
...
+   inlet: "L2LoadBalancer"
+   l2LoadBalancer:
+     addressPool: mypool
+     sourceRanges:
+       - 10.0.0.0/8
+   nodeSelector:
+     node-role.deckhouse.io/frontend: ""
+```
+
+По тем же соображениям что и в пердыдущем примере speaker-поды MetalLB должны быть запущены на тех же узлах, что и поды Ingress–контроллера.
```
Comment (apolovov) [inline modules/402-ingress-nginx/docs/EXAMPLES_RU.md:151]:
```suggestion
По тем же соображениям что, и в предыдущем примере, speaker-поды MetalLB должны быть запущены на тех же узлах, что и поды Ingress–контроллера.
```

File: modules/402-ingress-nginx/docs/EXAMPLES_RU.md [hunk 20]
```diff
@@ -128,4 +128,41 @@ metallb:
...
+
+Контроллер должен получать реальные IP-адреса клиентов — поэтому его Service создается с параметром `externalTrafficPolicy: Local` (запрещая межузловой SNAT), и для удовлетворения данного параметра MetalLB speaker анонсирует этот Service только с тех узлов, где запущены целевые поды.
```
Comment (apolovov) [inline modules/402-ingress-nginx/docs/EXAMPLES_RU.md:153]:
```suggestion
Контроллер должен знать реальные IP-адреса клиентов — поэтому его Service создается с параметром `externalTrafficPolicy: Local` (запрещая межузловой SNAT), и для удовлетворения данного параметра MetalLB speaker анонсирует этот Service только с тех узлов, где запущены целевые поды.
```

File: modules/402-ingress-nginx/docs/EXAMPLES_RU.md [hunk 21]
```diff
@@ -128,4 +128,41 @@ metallb:
...
+
+Таким образом, для данного примера [конфигурация модуля `metallb`](../380-metallb/configuration.html) должна быть такой:
+
+```yaml
+metallb:
```
Comment (apolovov) [inline modules/402-ingress-nginx/docs/EXAMPLES_RU.md:158]:
```suggestion
kind: ModuleConfig

spec:
  settings:
  ...
```

**Review by apolovov — CHANGES_REQUESTED** [review]

File: ee/modules/381-ltwo-load-balancer/crds/l2-load-balncer.yaml [hunk 22]
```diff
@@ -0,0 +1,120 @@
...
+        openAPIV3Schema:
+          type: object
+          required: [ 'spec' ]
+          properties:
+            spec:
+              type: object
+              properties:
+                default: { }
+                properties:
+                  nodeSelector:
```
Comment (apolovov) [inline ee/modules/381-ltwo-load-balancer/crds/l2-load-balncer.yaml:29]:
```suggestion
                  addressPool: <mandatory>
                  nodeSelector:
```

File: ee/modules/381-ltwo-load-balancer/templates/controller/ipaddresspools.yaml [hunk 23]
```diff
@@ -0,0 +1,14 @@
...
+apiVersion: metallb.io/v1beta1
+kind: IPAddressPool
+metadata:
+  name: {{ $ipaddressPool.name }}
+  namespace: d8-{{ $.Chart.Name }}
+  {{- include "helm_lib_module_labels" (list $ (dict "app" "controller")) | nindent 2 }}
+spec:
+  addresses:
+  {{- $ipaddressPool.addresses | toYaml | nindent 4 }}
+  autoAssign: {{ index $ipaddressPool "auto-assign" }}
```
[the remark this answers is not in the document]
Comment (apolovov) [reply ee/modules/381-ltwo-load-balancer/templates/controller/ipaddresspools.yaml:12]: and remove autoassign option from MC.

File: ee/modules/381-ltwo-load-balancer/templates/monitoring.yaml [hunk 24]
```diff
@@ -0,0 +1,2 @@
+{{- include "helm_lib_grafana_dashboard_definitions" . }}
```
Comment (apolovov) [inline ee/modules/381-ltwo-load-balancer/templates/monitoring.yaml:1]: TODO (not now, to think):
* make a module metallb-common, which is enabled if one of modules is enabled (mlb or l2lb).
* this modele deploys grafana dashboars and alerts and whatever.

File: ee/modules/381-ltwo-load-balancer/templates/services/service.yaml [hunk 25]
```diff
@@ -0,0 +1,31 @@
+{{- if .Values.ltwoLoadBalancer.internal }}
```
Comment (apolovov) [inline ee/modules/381-ltwo-load-balancer/templates/services/service.yaml:1]: services/service.yaml -> user-services/service.yaml

File: ee/modules/381-ltwo-load-balancer/templates/speaker/daemonset.yaml [hunk 26]
```diff
@@ -0,0 +1,154 @@
...
+        l2-load-balancer.network.deckhouse.io/member: ""
+
+      {{- include "helm_lib_tolerations" (tuple . "any-node") | nindent 6 }}
+
+      imagePullSecrets:
+        - name: deckhouse-registry
+      containers:
+        - args:
+            - --host=127.0.0.1
+            - --port=7472
```
Comment (apolovov) [inline ee/modules/381-ltwo-load-balancer/templates/speaker/daemonset.yaml:59]:
```suggestion
            - --port=7472
            - --lb-class=...
```

**Review by apolovov — CHANGES_REQUESTED** [review]

File: ee/modules/110-istio/hooks/ee/lib/crd/istiofederation.go [hunk 27]
```diff
@@ -8,7 +8,7 @@ package crd
 import metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 
 type IstioFederation struct {
-	metav1.TypeMeta `json:",inline"`
+	metav1.TypeMeta `json:",inline,omitempty"`
```
Comment (apolovov) [inline ee/modules/110-istio/hooks/ee/lib/crd/istiofederation.go:11]: ?

Comment (apolovov) [inline ee/modules/110-istio/hooks/ee/lib/crd/istiofederation.go:11]:
```suggestion
	metav1.TypeMeta `json:",inline"`
```

File: ee/modules/381-ltwo-load-balancer/crds/l2-load-balncer.yaml [hunk 28]
```diff
@@ -0,0 +1,135 @@
...
+              default: {}
+              properties:
+                nodeSelector:
+                  type: object
+                  additionalProperties:
+                    type: string
+                  description: |
+                    The same as in the Pods `spec.nodeSelector` parameter in Kubernetes.
+
+                    If parameter is omitted, L2 LoadBalancer will be placed on all nodes.
```
Comment (apolovov) [inline ee/modules/381-ltwo-load-balancer/crds/l2-load-balncer.yaml:37]:
```suggestion
                    If parameter is omitted, L2 LoadBalancer's speakers will be placed on all nodes.
```

File: ee/modules/381-ltwo-load-balancer/crds/l2-load-balncer.yaml [hunk 29]
```diff
@@ -0,0 +1,135 @@
...
+                  properties:
+                    addressPool:
+                      type: string
+                    sourceRanges:
+                      type: array
+                      default: []
+                      description: |
+                        IP ranges (CIDR) that are allowed to access the load balancer.
+
+                        The cloud provider may not support this option or ignore it.                       .
```
Comment (apolovov) [inline ee/modules/381-ltwo-load-balancer/crds/l2-load-balncer.yaml:50]:
```suggestion
```

File: ee/modules/381-ltwo-load-balancer/crds/l2-load-balncer.yaml [hunk 30]
```diff
@@ -0,0 +1,135 @@
...
+                    type: string
+                  description: |
+                    The same as in the Pods `spec.nodeSelector` parameter in Kubernetes.
+
+                    If parameter is omitted, L2 LoadBalancer will be placed on all nodes.
+                service:
+                  type: object
+                  required: ['addressPool', 'labelSelector', 'ports']
+                  properties:
+                    addressPool:
```
Comment (apolovov) [inline ee/modules/381-ltwo-load-balancer/crds/l2-load-balncer.yaml:42]: addressPool to the .spec.

File: ee/modules/381-ltwo-load-balancer/hooks/common.go [hunk 31]
```diff
@@ -0,0 +1,87 @@
...
+	Ports         []corev1.ServicePort  `json:"ports"`
+	Nodes         []map[string]string   `json:"nodes"`
+}
+
+type L2LoadBalancer struct {
+	metav1.TypeMeta `json:",inline"`
+	// Standard object's metadata.
+	// More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#metadata
+	// +optional
+	metav1.ObjectMeta `json:"metadata,omitempty"`
```
Comment (apolovov) [inline ee/modules/381-ltwo-load-balancer/hooks/common.go:32]: Figure out, how to use the json field.

File: ee/modules/381-ltwo-load-balancer/hooks/common.go [hunk 32]
```diff
@@ -0,0 +1,87 @@
...
+type L2LoadBalancerSpec struct {
+	NodeSelector map[string]string         `json:"nodeSelector"`
+	Service      L2LoadBalancerSpecService `json:"service"`
+}
+
+type L2LoadBalancerSpecService struct {
+	AddressPool   string                `json:"addressPool"`
+	LabelSelector *metav1.LabelSelector `json:"labelSelector"`
+	Ports         []corev1.ServicePort  `json:"ports"`
+	Nodes         []map[string]string   `json:"nodes"`
```
Comment (apolovov) [inline ee/modules/381-ltwo-load-balancer/hooks/common.go:46]: move Nodes to L2LoadBalancerSpec.

File: ee/modules/381-ltwo-load-balancer/hooks/common.go [hunk 33]
```diff
@@ -0,0 +1,87 @@
...
+	SourceRanges  []string              `json:"sourceRanges,omitempty"`
+}
+
+type NodeLabelsInfo struct {
+	Name      string
+	Labels    map[string]string
+	IsLabeled bool
```
Comment (apolovov) [inline ee/modules/381-ltwo-load-balancer/hooks/common.go:53]:
```suggestion
	IsLabeled bool // is there `l2-load-balancer.network.deckhouse.io/member` label
```

File: ee/modules/381-ltwo-load-balancer/hooks/common.go [hunk 34]
```diff
@@ -0,0 +1,87 @@
...
+
+type L2LoadBalancerSpecService struct {
+	AddressPool   string                `json:"addressPool"`
+	LabelSelector *metav1.LabelSelector `json:"labelSelector"`
+	Ports         []corev1.ServicePort  `json:"ports"`
+	Nodes         []map[string]string   `json:"nodes"`
+	SourceRanges  []string              `json:"sourceRanges,omitempty"`
+}
+
+type NodeLabelsInfo struct {
```
Comment (apolovov) [inline ee/modules/381-ltwo-load-balancer/hooks/common.go:50]:
```suggestion
type NodeInfo struct {
```

File: ee/modules/381-ltwo-load-balancer/hooks/handle_node_labels.go [hunk 35]
```diff
@@ -0,0 +1,122 @@
...
+		if key != memberLabelKey {
+			labels[key] = value
+		}
+	}
+
+	node.Labels = labels
+	return sdk.ToUnstructured(node)
+}
+
+func appendL2LabelsFromNode(obj *unstructured.Unstructured) (*unstructured.Unstructured, error) {
```
Comment (apolovov) [inline ee/modules/381-ltwo-load-balancer/hooks/handle_node_labels.go:98]:
```suggestion
func appendL2LabelsToNode(obj *unstructured.Unstructured) (*unstructured.Unstructured, error) {
```

File: ee/modules/381-ltwo-load-balancer/template_tests/module_test.go [hunk 36]
```diff
@@ -0,0 +1,120 @@
...
+  addresses:
+  - 192.168.0.0/24
+- name: mypool2
+  avoid-buggy-ips: false
+  addresses:
+  - 192.168.1.0-192.168.1.255
+`
+)
+
+var _ = Describe("Module :: ltwo-load-balancer :: helm template ::", func() {
```
Comment (apolovov) [inline ee/modules/381-ltwo-load-balancer/template_tests/module_test.go:35]:
```suggestion
var _ = Describe("Module :: l2-load-balancer :: helm template ::", func() {
```

File: ee/modules/381-ltwo-load-balancer/hooks/handle_node_labels.go [hunk 37]
```diff
@@ -0,0 +1,122 @@
+/*
```
[the remark this answers is not in the document]
Comment (apolovov) [reply ee/modules/381-ltwo-load-balancer/hooks/handle_node_labels.go:1]: The decision — merge both hooks to handle_load_balancers.go.

**Review by apolovov — CHANGES_REQUESTED** [review]

File: ee/modules/110-istio/hooks/ee/lib/crd/istiofederation.go (hunk 27 above)
Comment (Beastlex) [reply ee/modules/110-istio/hooks/ee/lib/crd/istiofederation.go:11]: omitempty is unnecessary

**Review by apolovov — CHANGES_REQUESTED** [review]

File: deckhouse-controller/cmd/deckhouse-controller/register-go-hooks.go [hunk 38]
```diff
@@ -21,7 +22,7 @@ import (
 	_ "github.com/deckhouse/deckhouse/ee/modules/160-multitenancy-manager/hooks/apis/deckhouse.io/v1alpha1"
 	_ "github.com/deckhouse/deckhouse/ee/modules/160-multitenancy-manager/hooks/apis/deckhouse.io/v1alpha2"
 	_ "github.com/deckhouse/deckhouse/ee/modules/300-prometheus/hooks"
-	_ "github.com/deckhouse/deckhouse/ee/modules/380-metallb/hooks"
+	_ "github.com/deckhouse/deckhouse/ee/modules/381-ltwo-load-balancer/hooks"
```
Comment (apolovov) [inline deckhouse-controller/cmd/deckhouse-controller/register-go-hooks.go:25]: l2 is fixed.

File: ee/modules/381-ltwo-load-balancer/monitoring/grafana-dashboards/kubernetes-cluster/l2lb-pool.json [hunk 39]
```diff
@@ -0,0 +1,300 @@
+{
```
Comment (apolovov) [inline ee/modules/381-ltwo-load-balancer/monitoring/grafana-dashboards/kubernetes-cluster/l2lb-pool.json:1]: pleas check jic.

File: ee/modules/402-ingress-nginx/templates/controller-l2/service-l2-load-balancer.yaml [hunk 40]
```diff
@@ -0,0 +1,37 @@
+{{- $context := . }}
```
Comment (apolovov) [inline ee/modules/402-ingress-nginx/templates/controller-l2/service-l2-load-balancer.yaml:1]: service-l2-load-balancer.yaml -> l2-load-balancer.yaml

File: modules/402-ingress-nginx/hooks/get_ingress_controllers.go [hunk 41]
```diff
@@ -29,6 +29,11 @@ type Controller struct {
 	Spec map[string]interface{} `json:"spec"`
 }
 
+type NodeLabelsInfo struct {
+	Name   string
+	Labels map[string]string
+}
+
```
Comment (apolovov) [inline modules/402-ingress-nginx/hooks/get_ingress_controllers.go:36]: Don't need the type here.

File: testing/openapi_validation/library.go [hunk 42]
```diff
@@ -36,6 +36,12 @@ const (
 	parserConcurrentCount = 50
 )
 
+var (
+	modulesWithSameCRDs = map[string]struct{}{
```
Comment (apolovov) [inline testing/openapi_validation/library.go:40]:
```suggestion
	modulesWithCRDEditions = map[string]struct{}{
```

File: testing/openapi_validation/validators/edition_crd_subsets.go [hunk 43]
```diff
@@ -0,0 +1,120 @@
+/*
+Copyright 2021 Flant JSC
```
Comment (apolovov) [inline testing/openapi_validation/validators/edition_crd_subsets.go:2]:
```suggestion
Copyright 2024 Flant JSC
```

File: ee/modules/381-l2-load-balancer/crds/l2-load-balncer.yaml [hunk 44]
```diff
@@ -0,0 +1,87 @@
...
+                  sourceRanges:
+                    type: array
+                    x-doc-d8Revision: ee
+                    default: []
+                    description: |
+                      IP ranges (CIDR) that are allowed to access the load balancer.
+                    items:
+                      type: string
+                      pattern: '^(([0-9]|[1-9][0-9]|1[0-9]{2}|2[0-4][0-9]|25[0-5])\.){3}([0-9]|[1-9][0-9]|1[0-9]{2}|2[0-4][0-9]|25[0-5])(\/(3[0-2]|[1-2][0-9]|[0-9]))$'
+                  labelSelector:
```
Comment (apolovov) [inline ee/modules/381-l2-load-balancer/crds/l2-load-balncer.yaml:56]:
```suggestion
                  selector:
```
!!!

**Review by apolovov — APPROVED** [review]

**Review by konstantin-axenov — APPROVED** [review]

**Review by nabokihms — APPROVED** [review]

**Pull request merged**